src/graphnet/models/components/layers.py

```diff
@@ -593,3 +602,393 @@ def forward(
+        self.rezero = rezero
+        self.deg_scaler = deg_scaler
+        self.activation = activation()
+
+        self.attention = GritSparseMHA(
+            in_dim=in_dim,
+            out_dim=out_dim // num_heads,
+            num_heads=num_heads,
+            use_bias=attn_bias,
+            dropout=attn_dropout,
+            clamp=attn_clamp,
+            activation=attn_activation,
+            edge_enhance=attn_edge_enhance,
+        )
+
+        self.fc1_x = nn.Linear(out_dim // num_heads * num_heads, out_dim)
+        if enable_edge_transform:
+            self.fc1_e = nn.Linear(out_dim // num_heads * num_heads, out_dim)
+        else:
+            self.fc1_e = nn.Identity()
+
+        if self.deg_scaler:
+            self.deg_coef = nn.Parameter(
+                torch.zeros(1, out_dim // num_heads * num_heads, 2)
+            )
+            nn.init.xavier_normal_(self.deg_coef)
+
+        if norm == nn.LayerNorm:
+            self.norm1_x = norm(out_dim)
+            self.norm1_e = self.norm(out_dim) if norm_edges else nn.Identity()
+        elif norm == nn.BatchNorm1d:
+            self.norm1_x = norm(
+                out_dim,
+                track_running_stats=self.batch_norm_runner,
+                eps=1e-5,
+                momentum=self.batch_norm_momentum,
+            )
+            self.norm1_e = (
+                norm(
+                    out_dim,
+                    track_running_stats=self.batch_norm_runner,
+                    eps=1e-5,
+                    momentum=self.batch_norm_momentum,
+                )
+                if norm_edges
+                else nn.Identity()
+            )
+        else:  # TODO: Maybe just set this to nn.Identity. -PW
+            raise ValueError(
+                "GritTransformerLayer normalization layer must be 'LayerNorm' \
+                    or 'BatchNorm1d'!"
+            )
+
+        # FFN for x
+        self.FFN_x_layer1 = nn.Linear(out_dim, out_dim * 2)
+        self.FFN_x_layer2 = nn.Linear(out_dim * 2, out_dim)
+
+        if norm == nn.LayerNorm:
+            self.norm2_x = norm(out_dim)
+        elif norm == nn.BatchNorm1d:
+            self.norm2_x = norm(
+                out_dim,
+                track_running_stats=self.batch_norm_runner,
+                eps=1e-5,
+                momentum=self.batch_norm_momentum,
+            )
+
+        if self.rezero:  # Learnable scaling parameters
+            self.alpha1_x = nn.Parameter(torch.zeros(1, 1))
+            self.alpha2_x = nn.Parameter(torch.zeros(1, 1))
+            self.alpha1_e = nn.Parameter(torch.zeros(1, 1))
+
+        self.dropout1 = nn.Dropout(dropout)  # Post-attention dropout on x
+        self.dropout2 = nn.Dropout(dropout)  # Post-attention dropout on e
+        self.dropout3 = nn.Dropout(dropout)  # Post-FFN dropout on x
+
+    def forward(self, data: Data) -> Data:
+        """Forward pass."""
+        x = data.x
+        num_nodes = data.num_nodes
+        log_deg = get_log_deg(data)
```

Comment:
Was there some particular reason why you needed the utility function to grab or calculate this quantity?

Naively I would've thought the degree could've been calculated directly in the forward pass like  so: 

```python
from torch_geometric.utils import degree
log_deg = torch.log10(degree(data.edge_index[0]) + 1)
```